src/Framework/PaymentGateways/PaymentGateway.php

```diff
@@ -214,63 +226,45 @@ public function handleGatewaySubscriptionCommand(
         );
     }
 
-    /**
-     * Handle gateway route method
-     *
-     * @param  int  $donationId
-     * @param  string  $method
-     *
-     * @since 2.18.0
-     *
-     * @return void
-     */
-    public function handleGatewayRouteMethod($donationId, $method)
-    {
-        try {
-            $this->handleResponse( $this->$method( $donationId ) );
-        } catch (PaymentGatewayException $paymentGatewayException) {
-            $this->handleResponse(response()->json($paymentGatewayException->getMessage()));
-        } catch (Exception $exception) {
-            PaymentGatewayLog::error($exception->getMessage());
-            $this->handleResponse(response()->json(
-                __( 'An unexpected error occurred while processing your donation.  Please try again or contact us to help resolve.', 'give' )
-            ));
-        }
-    }
-
     /**
      * Generate gateway route url
      *
      * @since 2.18.0
+     * @unreleased remove $donationId param in favor of args
      *
      * @param  string  $gatewayMethod
-     * @param  int  $donationId
      * @param  array|null  $args
      *
      * @return string
+     *
      */
-    public function generateGatewayRouteUrl($gatewayMethod, $donationId, $args = null)
+    public function generateGatewayRouteUrl($gatewayMethod, $args = null)
     {
-        return Call::invoke(GenerateGatewayRouteUrl::class, $this->getId(), $gatewayMethod, $donationId, $args);
+        return Call::invoke(GenerateGatewayRouteUrl::class, $this->getId(), $gatewayMethod, $args);
     }
 
-
     /**
-     * Handle Response
+     * Generate secure gateway route url
      *
-     * @since 2.18.0
+     * @unreleased
+     *
+     * @param  string  $gatewayMethod
+     * @param  array|null  $args
+     *
+     * @return string
      *
-     * @param  RedirectResponse|JsonResponse  $type
      */
-    public function handleResponse($type)
+    public function generateSecureGatewayRouteUrl($gatewayMethod, $args = null)
     {
-        if ($type instanceof RedirectResponse) {
-            wp_redirect($type->getTargetUrl());
-            exit;
-        }
-
-        if ($type instanceof JsonResponse) {
-            wp_send_json(['data' => $type->getData()]);
-        }
+        $nonce = RouteSignature::make($this->getId(), $gatewayMethod, $args);
+
+        return Call::invoke(
+            GenerateGatewayRouteUrl::class,
+            $this->getId(),
+            $gatewayMethod,
+            array_merge($args, [
```

Comment:
@jonwaldstein this will generate a PHP warning because default value of `$args` is `null` but function expect `array`